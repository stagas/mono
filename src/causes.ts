import { ParserErrorCauses } from 'tinypratt'

export class CompilerErrorCause extends ParserErrorCauses.ParserErrorCause {
  name = 'CompilerUnknownError'
}
export class ReferenceErrorCause extends CompilerErrorCause {
  name = 'CompilerReferenceError'
}
export class TypeErrorCause extends CompilerErrorCause {
  name = 'CompilerTypeError'
}
export class SyntaxErrorCause extends CompilerErrorCause {
  name = 'CompilerSyntaxError'
}
export class InvalidErrorCause extends CompilerErrorCause {
  name = 'CompilerInvalidError'
}
