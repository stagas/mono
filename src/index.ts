export * from './compiler'
export * from './get-shared-worker-port'
export * from './linker-service'
export * from './parser'
export * from './syntax'
