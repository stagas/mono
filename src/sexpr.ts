export interface SExpr extends Array<string | SExpr> {
  [n: number]: string | SExpr
}

export const S = (p: string | SExpr, x = 0): string => {
  return Array.isArray(p)
    ? `${' '.repeat(x)}(${p[0]} ${p
        .slice(1)
        .map(e => (Array.isArray(e) ? '\n' : '') + S(e, x + 2))
        .join(' ')})`
    : p
}

export const S0 = (p: string | SExpr): string => {
  return Array.isArray(p)
    ? `(${p[0]} ${p
        .slice(1)
        .map(e => S0(e))
        .join(' ')})`
    : p
}
