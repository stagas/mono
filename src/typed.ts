import { SExpr } from './sexpr'
import { Token } from './parser'

export enum Type {
  any = 'any',
  bool = 'bool',
  i32 = 'i32',
  f32 = 'f32',
}

export const Types: Type[] = [Type.any, Type.bool, Type.i32, Type.f32]

export const Typed = (panic: (s: string, t: string) => string) => {
  const types = new Map<object | string, Type>()

  const OpTypeCast: Record<Type, Partial<Record<Type, string>>> = {
    [Type.any]: {},
    [Type.f32]: {
      [Type.i32]: 'f32.convert_i32_u',
      [Type.bool]: 'f32.convert_i32_s',
    },
    [Type.i32]: {
      [Type.f32]: 'i32.trunc_f32_u',
    },
    [Type.bool]: {
      [Type.f32]: 'i32.trunc_f32_s',
    },
  }

  /** looks up and returns the type of `x`, if found, otherwise returns type `any` */
  const typeOf = (x: undefined | string | SExpr): Type => ((x && types.get(x)) ?? Type.any) as Type

  /** marks sexpr `x` to be of type `type` */
  const typeAs = (type: Type, x: SExpr) => (types.set(x, type), x)

  /** creates a cast operation if the given value `x` doesn't satisfy `type` */
  const cast = (type: Type, x: SExpr) => {
    const childType = typeOf(x)
    if (childType != type) {
      const castOp = OpTypeCast[type][childType]
      if (!castOp) return typeAs(type, x) // noop cast, but change the type for x
      return typeAs(type, [castOp, x])
    } else {
      return typeAs(type, x) // x is any or unknown, so set the type for x
    }
  }

  /** returns the highest precision type of the given values */
  const hi = (...values: SExpr): Type => {
    const weights = values.map(x => Types.indexOf(typeOf(x)))
    return Types[Math.max(...weights)]
  }

  /** returns at least the precision of the given type among the given types */
  const max = (type: Type, ...types: Type[]): Type => {
    return Types[Math.max(Types.indexOf(type), ...types.map(x => Types.indexOf(x)))]
  }

  /** types a wasm operation with the correct prefix (f32 or i32) */
  const top = (type: Type, ops: SExpr): SExpr => {
    const prefix = type == Type.f32 ? 'f32' : 'i32'
    let [op, ...children] = ops // eslint-disable-line prefer-const
    children = children.map(x => cast(type, x as SExpr))
    return typeAs(type, [prefix + '.' + op, ...children])
  }

  /** infers the type of a token literal string: bool for 0 or 1, i32 for integers and f32 for floats */
  const infer = (x: Token): Type => {
    if (x == '0' || x == '1') return Type.bool
    else if (!x.includes('.')) return Type.i32
    else if (x.includes('.')) return Type.f32
    else throw new TypeError(panic('cannot infer type for', x))
  }

  return { typeOf, typeAs, cast, hi, max, top, infer }
}
