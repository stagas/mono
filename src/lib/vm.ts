import type { LinkerConfig } from '../linker'
import { Type, W, max } from '../typed'
import { Arg } from '..'

export const env = (config: LinkerConfig) => `;;wasm
  (import "env" "memory"
    (memory
      ${config.memory.initial}
      ${config.memory.maximum}
      ${config.memory.shared ? 'shared' : ''}
    )
  )

  (global $output_mem_ptr (i32) (i32.const 0))
  (global $input_mem_ptr (i32) (i32.const 0x40000))
  (global $global_mem_ptr (export "global_mem_ptr") (mut i32) (i32.const 0x80000))
  (global $sr (export "sampleRate") (mut f32) (f32.const 44100.0))
  (global $t (export "currentTime") (mut f32) (f32.const 0))
  (global $pi (f32) (f32.const 3.1415927410125732))
  (global $pi2 (f32) (f32.const 6.2831854820251465))
`

// TODO: length should be the global_buffer_length

export interface VM {
  fill(frame: number, length: number, ...args: number[]): void
}

export const fill = ({ type, params }: { type: Type; params: Arg[] }) => `;;wasm
  (func $fill (export "fill")
    ;; song position starting frame
    (param $frame i32)
    ;; buffer length to fill
    (param $length i32)
    ;; ...params
    ${params.map(x => `(param $${x.id} ${max(Type.i32, x.type!)})`).join(' ')}

    ;; buffer index position
    (local $index i32)

    ;; index = 0
    (local.set $index (i32.const 0))

    ;; do
    (loop $loop
      (global.set $global_mem_ptr (i32.const 0x80000))

      ;; run initializers
      (call $__begin__)

      ;; t = $frame / $sr
      (global.set $t
        (f32.div
          (f32.convert_i32_u (local.get $frame))
          (global.get $sr)
        )
      )

      ;; buffer[$index] =
      (f32.store (i32.shl (local.get $index) (i32.const 2))
        ;; f(...params)
        ${W(type) < W(Type.f32) ? '(f32.convert_i32_s ' : ''}
        (call $f ${params.map(x => `(local.get $${x.id})`).join(' ')})
        ${W(type) < W(Type.f32) ? ')' : ''}
      )

      ;; $index++
      (local.set $index (i32.add (local.get $index) (i32.const 1)))

      ;; $frame++
      (local.set $frame (i32.add (local.get $frame) (i32.const 1)))

      ;; if ($index !== $length) continue $loop
      (br_if $loop (i32.ne (local.get $index) (local.get $length)))
    )
  )
`
