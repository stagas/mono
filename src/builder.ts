import rfdc from 'rfdc'
import make, * as wat from 'wat-compiler'
import { Imports, Module, compile } from './compiler'
import * as lib from './lib.wat'
import { parse } from './parser'

// eslint-disable-next-line @typescript-eslint/no-explicit-any
let modlib: any
const copy = rfdc({ proto: true, circles: false })
const imports: Imports = {}

export const build = (s: string, scope = {}, extraLib = {}, extraRun = (_mod: Module) => []) => {
  console.time('build')

  if (!modlib) {
    console.time('lib')
    const libsrc = Object.values(lib).join('\n') + '\n' + Object.values(extraLib).join('\n')
    modlib = wat.compile(wat.parse(wat.tokenize('(module ' + libsrc + ')')))
    for (const code of modlib.module.codes) {
      const [params, result] = modlib.module.types[code.type_idx].split(',')
      imports[code.name] = { params: params.split(' '), result: result.split(' ')[0] }
    }
    console.timeEnd('lib')
  }

  // console.time('parse')
  const ast = parse(s)
  // console.timeEnd('parse')

  // console.time('compile')
  const mod = compile(ast, scope, imports)
  // console.timeEnd('compile')

  // console.time('make')
  const buffer = make(mod.toString(extraRun(mod)), { metrics: false }, copy(modlib))
  // console.timeEnd('make')

  console.timeEnd('build')

  return {
    module: mod,
    buffer,
  }
}
