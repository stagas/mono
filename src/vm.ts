import { Token } from 'lexer-next'
import { Arg } from './compiler'
import { Module } from './compiler'
import * as libmono from './lib/mono'
import * as libvm from './lib/vm'
import * as libwat from './lib/wat'
import { Linker, LinkerConfig } from './linker'
import { Type } from './typed'

const defaultConfig: LinkerConfig = {
  memory: {
    initial: 32, // TODO: autocalculate page size
    maximum: 32,
  },
  metrics: false,
}

const parseArg =
  /(?<arg>\.?\s*(?<id>[a-zA-Z_$][a-zA-Z0-9_$]*)(?<range>\[[^\]]*?\])?(=.*?(?<default>[^\s,)]+)?)?).*?(,|\)\s*=)/s

export class MonoParam {
  id!: Token
  sourceIndex!: number
  source!: {
    arg: string
    id: string
    range: string
    default: string
  }
  name!: string
  minValue!: number
  maxValue!: number
  defaultValue!: number
  normalValue!: number
  scaleValue!: number

  constructor(data: Partial<MonoParam>) {
    Object.assign(this, data)
    this.scaleValue = this.maxValue - this.minValue
    this.normalValue = this.normalize(this.defaultValue)
  }

  normalize(value: number) {
    return (value - this.minValue) / this.scaleValue
  }

  scale(normal: number) {
    return normal * this.scaleValue + this.minValue
  }
}

export class VM {
  isReady = false
  skipMono = false

  code = ''

  config: LinkerConfig

  linker: Linker
  module?: Module
  instance?: WebAssembly.Instance

  inputs: Float32Array[]
  outputs: Float32Array[]

  constructor(config: Partial<LinkerConfig> = {}) {
    this.config = { ...defaultConfig, ...config }
    this.linker = new Linker(this.config)
    this.inputs = [new Float32Array(this.linker.memory.buffer, 0x40000, 0x10000)] // TODO: config size
    this.outputs = [new Float32Array(this.linker.memory.buffer, 0, 0x10000)]
    this.linkLib()
  }

  linkLib() {
    this.linker.linkLib({ env: libvm.env, ...libwat })
    if (!this.skipMono) this.linker.linkMono(libmono)
    this.linker.linkSExpr(() => [libvm.fill({ type: this.f_type, params: this.f_params })])
  }

  get exports() {
    return this.instance?.exports as {
      [k: string]: ((...args: number[]) => number) | WebAssembly.Global | any
    } & libvm.VM
  }

  get funcs() {
    return this.linker.module!.funcs
  }

  get params(): MonoParam[] {
    return Object.values(this.module?.funcs ?? {})
      .map(f => [f.id, f.params.filter(p => p.export)])
      .filter(x => x[1].length)
      .map(
        ([id, params]: any) =>
          params.map((p: Arg) => {
            const export_id = p.id.as('export/' + id + '/' + p.id) as Token
            const export_id_min = p.id.as('export/' + id + '/' + p.id + '/min') as Token
            const export_id_max = p.id.as('export/' + id + '/' + p.id + '/max') as Token

            const minValue = (this.instance!.exports[export_id_min] as any)?.value ?? 0
            const maxValue = (this.instance!.exports[export_id_max] as any)?.value ?? 1
            const defaultValue = (this.instance!.exports[export_id] as any)?.value
              ?? (maxValue - minValue) * 0.5 + minValue

            // find the index of the dot `.` that made the param export by moving
            // backwards from the id index that we have access to here
            const code = p.id.source.input
            const sourceIndex = code.lastIndexOf('.', export_id.source.index)
            const source = code.slice(sourceIndex).match(parseArg)!.groups as MonoParam['source']

            return new MonoParam({
              id: export_id,
              sourceIndex,
              source,
              name: export_id.toString(),
              minValue,
              maxValue,
              defaultValue,
            })
          })
        // argToMonoParam({ ...p, default: p.originalDefault, id: p.id.as('export/' + id + '/' + p.id) as Token })
        // )
      )
      .flat()
  }

  get f_type(): Type {
    return this.linker.module!.typeOf(this.funcs.f?.body)
  }

  get f_params(): Arg[] {
    return this.funcs.f?.params ?? [] // .map(argToMonoParam)
  }

  async setCode(code: string) {
    this.linker.link(code)
    await this.linker.make()

    // update code only after linker has been successful
    this.code = code

    // we copy these because they are instantiated at two async steps
    // before the linker is ready, and "module" and "instance" can diverge
    // if we were to access them directly from the linker, leading to errors
    this.module = this.linker.module
    this.instance = this.linker.instance

    this.isReady = true
    this.config.metrics && console.log('mono: ready')
  }
}
