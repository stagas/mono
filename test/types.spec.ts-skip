// @env browser
import { S } from '../src/sexpr'
import { Type } from '../src/typed'
import { VM } from '../src/vm'

const run = async (code: string) => {
  const vm = new VM()
  vm.skipMono = true
  await vm.setCode(code)
  let a_type, a_source
  if (vm.linker.module!.funcs.a) {
    a_type = vm.linker.module!.typeOf(vm.linker.module!.funcs.a.body!.at(-1))
    a_source = S(vm.linker.module!.funcs.a.source!)
  }
  return {
    vm,
    typeOf: vm.linker.module!.typeOf,
    funcs: vm.linker.module!.funcs,
    a_type,
    a_source,
    start_type: vm.linker.module!.typeOf(vm.linker.module!.funcs.__start__.body!.at(-1)),
    start_source: S(vm.linker.module!.funcs.__start__.source!),
    f_type: vm.linker.module!.typeOf(vm.linker.module!.funcs.f.body!.at(-1)),
    f_source: S(vm.linker.module!.funcs.f.source!),
    f: (...params: any) => {
      // simulate start of loop
      vm.exports.global_mem_ptr.value = 0
      return vm.exports.f.call(this, ...params)
    },
  }
}

describe('Types', () => {
  describe('const', () => {
    it('f32', async () => {
      {
        const m = await run('f()=1.5')
        expect(m.f_source).toMatchSnapshot()
        expect(m.f_type).toBe(Type.f32)
        expect(m.f()).toBe(1.5)
      }

      {
        const m = await run('f()=-1.5')
        expect(m.f_source).toMatchSnapshot()
        expect(m.f_type).toBe(Type.f32)
        expect(m.f()).toBe(-1.5)
      }
    })

    it('bool', async () => {
      {
        const m = await run('f()=0')
        expect(m.f_source).toMatchSnapshot()
        expect(m.f_type).toBe(Type.bool)
        expect(m.f()).toBe(0)
      }
      {
        const m = await run('f()=1')
        expect(m.f_source).toMatchSnapshot()
        expect(m.f_type).toBe(Type.bool)
        expect(m.f()).toBe(1)
      }
    })

    it('i32', async () => {
      {
        const m = await run('f()=2')
        expect(m.f_source).toMatchSnapshot()
        expect(m.f_type).toBe(Type.i32)
        expect(m.f()).toBe(2)
      }
      {
        const m = await run('f()=-2')
        expect(m.f_source).toMatchSnapshot()
        expect(m.f_type).toBe(Type.i32)
        expect(m.f()).toBe(-2)
      }
    })
  })

  describe('calling function', () => {
    it('i32', async () => {
      const m = await run('a()=2;f()=a()')
      expect(m.a_source).toMatchSnapshot()
      expect(m.a_type).toBe(Type.i32)
      expect(m.f_source).toMatchSnapshot()
      expect(m.f_type).toBe(Type.i32)
      expect(m.f()).toBe(2)
    })

    it('one level deep i32', async () => {
      const m = await run('b()=2;a()=b();f()=a()')
      expect(m.a_source).toMatchSnapshot()
      expect(m.a_type).toBe(Type.i32)
      expect(m.f_source).toMatchSnapshot()
      expect(m.f_type).toBe(Type.i32)
      expect(m.f()).toBe(2)
    })

    it('f32', async () => {
      const m = await run('a()=1.5;f()=a()')
      expect(m.a_source).toMatchSnapshot()
      expect(m.a_type).toBe(Type.f32)
      expect(m.f_source).toMatchSnapshot()
      expect(m.f_type).toBe(Type.f32)
      expect(m.f()).toBe(1.5)
    })

    it('one level deep f32', async () => {
      const m = await run('b()=1.5;a()=b();f()=a()')
      expect(m.a_source).toMatchSnapshot()
      expect(m.a_type).toBe(Type.f32)
      expect(m.f_source).toMatchSnapshot()
      expect(m.f_type).toBe(Type.f32)
      expect(m.f()).toBe(1.5)
    })

    it('switching i32 -> f32', async () => {
      const m = await run('a()=2;f()=a()+0.5')
      expect(m.a_source).toMatchSnapshot()
      expect(m.a_type).toBe(Type.i32)
      expect(m.f_source).toMatchSnapshot()
      expect(m.f_type).toBe(Type.f32)
      expect(m.f()).toBe(2.5)
    })

    it('switching f32 -> bool(i32)', async () => {
      const m = await run('a()=2.5;f()=!a()')
      expect(m.a_source).toMatchSnapshot()
      expect(m.a_type).toBe(Type.f32)
      expect(m.f_source).toMatchSnapshot()
      expect(m.f_type).toBe(Type.bool)
      expect(m.f()).toBe(0)
    })

    it('switching f32 -> bool(i32) -> i32', async () => {
      const m = await run('b()=2.5;a()=!b();f()=a()+5')
      expect(m.a_source).toMatchSnapshot()
      expect(m.a_type).toBe(Type.bool)
      expect(m.f_source).toMatchSnapshot()
      expect(m.f_type).toBe(Type.i32)
      expect(m.f()).toBe(5)
    })
  })

  describe('arguments', () => {
    it('f32', async () => {
      {
        const m = await run('f(x=1.0)=x')
        expect(m.f_source).toMatchSnapshot()
        expect(m.f_type).toBe(Type.f32)
        expect(m.f(1.5)).toBe(1.5)
      }

      {
        const m = await run('f(x=-1.0)=x')
        expect(m.f_source).toMatchSnapshot()
        expect(m.f_type).toBe(Type.f32)
        expect(m.f(1.5)).toBe(1.5)
      }

      {
        const m = await run('f(.x=-1.0)=x')
        expect(m.f_source).toMatchSnapshot()
        expect(m.f_type).toBe(Type.f32)
        expect(m.f(1.5)).toBe(1.5)
      }
    })

    it('i32', async () => {
      {
        const m = await run('f(x=2)=x')
        expect(m.f_source).toMatchSnapshot()
        expect(m.f_type).toBe(Type.i32)
        expect(m.f(1.5)).toBe(1)
      }

      {
        const m = await run('f(x=-2)=x')
        expect(m.f_source).toMatchSnapshot()
        expect(m.f_type).toBe(Type.i32)
        expect(m.f(1.5)).toBe(1)
      }

      {
        const m = await run('f(.x=-2)=x')
        expect(m.f_source).toMatchSnapshot()
        expect(m.f_type).toBe(Type.i32)
        expect(m.f(1.5)).toBe(1)
      }
    })

    it('bool', async () => {
      {
        const m = await run('f(x=1)=x')
        expect(m.f_source).toMatchSnapshot()
        expect(m.f_type).toBe(Type.bool)
        expect(m.f(1.5)).toBe(1)
      }
    })

    it('exported bool', async () => {
      {
        const m = await run('f(.x=1)=x')
        expect(m.f_source).toMatchSnapshot()
        expect(m.f_type).toBe(Type.bool)
        expect(m.f(1.5)).toBe(1)
      }
    })

    it('export no type use f32', async () => {
      {
        const m = await run('f(.x)=x')
        expect(m.f_source).toMatchSnapshot()
        expect(m.f_type).toBe(Type.f32)
        expect(m.f(1.5)).toBe(1.5)
      }
    })

    it('infer f32', async () => {
      {
        const m = await run('f(x)=x')
        expect(m.f_source).toMatchSnapshot()
        expect(m.f_type).toBe(Type.f32)
        expect(m.f(1.5)).toBe(1.5)
      }
    })

    it('from range f32', async () => {
      {
        const m = await run('f(x[1.0..2.0])=x')
        expect(m.f_source).toMatchSnapshot()
        expect(m.f_type).toBe(Type.f32)
        expect(m.f(1.5)).toBe(1.5)
      }
    })

    it('from range i32', async () => {
      {
        const m = await run('f(x[1..2])=x')
        expect(m.f_source).toMatchSnapshot()
        expect(m.f_type).toBe(Type.i32)
        expect(m.f(1.5)).toBe(1)
      }
    })

    it('exported from range i32', async () => {
      {
        const m = await run('f(.x[1..2])=x')
        expect(m.f_source).toMatchSnapshot()
        expect(m.f_type).toBe(Type.i32)
        expect(m.f(1.5)).toBe(1)
      }
    })

    it('from range bool', async () => {
      {
        const m = await run('f(x[0..1])=x')
        expect(m.f_source).toMatchSnapshot()
        expect(m.f_type).toBe(Type.bool)
        expect(m.f(1.5)).toBe(1)
      }
    })

    it('range i32..i32, default f32', async () => {
      {
        const m = await run('f(x[1..2]=3.0)=x')
        expect(m.f_source).toMatchSnapshot()
        expect(m.f_type).toBe(Type.f32)
        expect(m.funcs.f.params[0].range![0]).toMatchSnapshot()
        expect(m.funcs.f.params[0].range![1]).toMatchSnapshot()
        expect(m.f(1.5)).toBe(1.5)
      }
    })

    it('range i32..f32, default f32 = f32', async () => {
      {
        const m = await run('f(x[1..2.0]=3.0)=x')
        expect(m.f_source).toMatchSnapshot()
        expect(m.f_type).toBe(Type.f32)
        expect(m.funcs.f.params[0].range![0]).toMatchSnapshot()
        expect(m.funcs.f.params[0].range![1]).toMatchSnapshot()
        expect(m.f(1.5)).toBe(1.5)
      }
    })

    it('range f32..f32, default i32 = f32', async () => {
      {
        const m = await run('f(x[1.0..2.0]=3)=x')
        expect(m.f_source).toMatchSnapshot()
        expect(m.f_type).toBe(Type.f32)
        expect(m.funcs.f.params[0].range![0]).toMatchSnapshot()
        expect(m.funcs.f.params[0].range![1]).toMatchSnapshot()
        expect(m.funcs.f.params[0].default).toMatchSnapshot()
        expect(m.f(1.5)).toBe(1.5)
      }
    })

    it('range i32..i32, default i32 = i32', async () => {
      {
        const m = await run('f(x[2..5]=3)=x')
        expect(m.f_source).toMatchSnapshot()
        expect(m.f_type).toBe(Type.i32)
        expect(m.funcs.f.params[0].range![0]).toMatchSnapshot()
        expect(m.funcs.f.params[0].range![1]).toMatchSnapshot()
        expect(m.funcs.f.params[0].default).toMatchSnapshot()
        expect(m.f(1.5)).toBe(1)
      }
    })

    it('use default i32', async () => {
      {
        const m = await run('a(x=5)=x;f()=a()')
        expect(m.f_source).toMatchSnapshot()
        expect(m.f_type).toBe(Type.i32)
        expect(m.f()).toBe(5)
      }
    })

    it('use default exported i32', async () => {
      {
        const m = await run('a(.x=5)=x;f()=a()')
        expect(m.start_type).toMatchSnapshot()
        expect(m.start_source).toMatchSnapshot()
        expect(m.f_source).toMatchSnapshot()
        expect(m.f_type).toBe(Type.i32)
        expect(m.f()).toBe(5)
      }
    })

    it('use default f32', async () => {
      {
        const m = await run('a(x=5.5)=x;f()=a()')
        expect(m.f_source).toMatchSnapshot()
        expect(m.f_type).toBe(Type.f32)
        expect(m.f()).toBe(5.5)
      }
    })

    it('default to range i32', async () => {
      {
        const m = await run('a(x[1..5])=x;f()=a()')
        expect(m.f_source).toMatchSnapshot()
        expect(m.f_type).toBe(Type.i32)
        expect(m.f()).toBe(1)
      }
    })

    it('default to range bool', async () => {
      {
        const m = await run('a(x[0..1])=x;f()=a()')
        expect(m.f_source).toMatchSnapshot()
        expect(m.f_type).toBe(Type.bool)
        expect(m.f()).toBe(0)
      }
    })

    it('default to range f32', async () => {
      {
        const m = await run('a(x[0.5..1])=x;f()=a()')
        expect(m.f_source).toMatchSnapshot()
        expect(m.f_type).toBe(Type.f32)
        expect(m.f()).toBe(0.5)
      }
    })
  })
})
