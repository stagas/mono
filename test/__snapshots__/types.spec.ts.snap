
exports[`Types const f32 1`] = `
"(func $f 
  (export \\"f\\") 
  (result f32) 
  (f32.const 1.5))"
`;

exports[`Types const f32 2`] = `
"(func $f 
  (export \\"f\\") 
  (result f32) 
  (f32.mul 
    (f32.const -1) 
    (f32.const 1.5)))"
`;

exports[`Types const bool 1`] = `
"(func $f 
  (export \\"f\\") 
  (result i32) 
  (i32.const 0))"
`;

exports[`Types const bool 2`] = `
"(func $f 
  (export \\"f\\") 
  (result i32) 
  (i32.const 1))"
`;

exports[`Types const i32 1`] = `
"(func $f 
  (export \\"f\\") 
  (result i32) 
  (i32.const 2))"
`;

exports[`Types const i32 2`] = `
"(func $f 
  (export \\"f\\") 
  (result i32) 
  (i32.mul 
    (i32.const -1) 
    (i32.const 2)))"
`;

exports[`Types calling function i32 1`] = `
"(func $a 
  (export \\"a\\") 
  (result i32) 
  (i32.const 2))"
`;

exports[`Types calling function i32 2`] = `
"(func $f 
  (export \\"f\\") 
  (result i32) 
  (call $a))"
`;

exports[`Types calling function one level deep i32 1`] = `
"(func $a 
  (export \\"a\\") 
  (result i32) 
  (call $b))"
`;

exports[`Types calling function one level deep i32 2`] = `
"(func $f 
  (export \\"f\\") 
  (result i32) 
  (call $a))"
`;

exports[`Types calling function f32 1`] = `
"(func $a 
  (export \\"a\\") 
  (result f32) 
  (f32.const 1.5))"
`;

exports[`Types calling function f32 2`] = `
"(func $f 
  (export \\"f\\") 
  (result f32) 
  (call $a))"
`;

exports[`Types calling function one level deep f32 1`] = `
"(func $a 
  (export \\"a\\") 
  (result f32) 
  (call $b))"
`;

exports[`Types calling function one level deep f32 2`] = `
"(func $f 
  (export \\"f\\") 
  (result f32) 
  (call $a))"
`;

exports[`Types calling function switching i32 -> f32 1`] = `
"(func $a 
  (export \\"a\\") 
  (result i32) 
  (i32.const 2))"
`;

exports[`Types calling function switching i32 -> f32 2`] = `
"(func $f 
  (export \\"f\\") 
  (result f32) 
  (f32.add 
    (f32.convert_i32_s 
      (call $a)) 
    (f32.const 0.5)))"
`;

exports[`Types calling function switching f32 -> bool(i32) 1`] = `
"(func $a 
  (export \\"a\\") 
  (result f32) 
  (f32.const 2.5))"
`;

exports[`Types calling function switching f32 -> bool(i32) 2`] = `
"(func $f 
  (export \\"f\\") 
  (result i32) 
  (i32.eqz 
    (i32.trunc_f32_u 
      (call $a))))"
`;

exports[`Types calling function switching f32 -> bool(i32) -> i32 1`] = `
"(func $a 
  (export \\"a\\") 
  (result i32) 
  (i32.eqz 
    (i32.trunc_f32_u 
      (call $b))))"
`;

exports[`Types calling function switching f32 -> bool(i32) -> i32 2`] = `
"(func $f 
  (export \\"f\\") 
  (result i32) 
  (i32.add 
    (call $a) 
    (i32.const 5)))"
`;

exports[`Types arguments f32 1`] = `
"(func $f 
  (export \\"f\\") 
  (param $x f32) 
  (result f32) 
  (local.get $x))"
`;

exports[`Types arguments f32 2`] = `
"(func $f 
  (export \\"f\\") 
  (param $x f32) 
  (result f32) 
  (local.get $x))"
`;

exports[`Types arguments f32 3`] = `
"(func $f 
  (export \\"f\\") 
  (param $x f32) 
  (result f32) 
  (local.get $x))"
`;

exports[`Types arguments i32 1`] = `
"(func $f 
  (export \\"f\\") 
  (param $x i32) 
  (result i32) 
  (local.get $x))"
`;

exports[`Types arguments i32 2`] = `
"(func $f 
  (export \\"f\\") 
  (param $x i32) 
  (result i32) 
  (local.get $x))"
`;

exports[`Types arguments i32 3`] = `
"(func $f 
  (export \\"f\\") 
  (param $x i32) 
  (result i32) 
  (local.get $x))"
`;

exports[`Types arguments bool 1`] = `
"(func $f 
  (export \\"f\\") 
  (param $x i32) 
  (result i32) 
  (local.get $x))"
`;

exports[`Types arguments exported bool 1`] = `
"(func $f 
  (export \\"f\\") 
  (param $x i32) 
  (result i32) 
  (local.get $x))"
`;

exports[`Types arguments export no type use f32 1`] = `
"(func $f 
  (export \\"f\\") 
  (param $x f32) 
  (result f32) 
  (local.get $x))"
`;

exports[`Types arguments infer f32 1`] = `
"(func $f 
  (export \\"f\\") 
  (param $x f32) 
  (result f32) 
  (local.get $x))"
`;

exports[`Types arguments from range f32 1`] = `
"(func $f 
  (export \\"f\\") 
  (param $x f32) 
  (result f32) 
  (local.get $x))"
`;

exports[`Types arguments from range i32 1`] = `
"(func $f 
  (export \\"f\\") 
  (param $x i32) 
  (result i32) 
  (local.get $x))"
`;

exports[`Types arguments exported from range i32 1`] = `
"(func $f 
  (export \\"f\\") 
  (param $x i32) 
  (result i32) 
  (local.get $x))"
`;

exports[`Types arguments from range bool 1`] = `
"(func $f 
  (export \\"f\\") 
  (param $x i32) 
  (result i32) 
  (local.get $x))"
`;

exports[`Types arguments range i32..i32, default f32 1`] = `
"(func $f 
  (export \\"f\\") 
  (param $x f32) 
  (result f32) 
  (local.get $x))"
`;

exports[`Types arguments range i32..i32, default f32 2`] = `
Array [
  "f32.convert_i32_s",
  Array [
    "i32.const",
    Token {
      "0": "1",
      "group": "num",
    },
  ],
]
`;

exports[`Types arguments range i32..i32, default f32 3`] = `
Array [
  "f32.convert_i32_s",
  Array [
    "i32.const",
    Token {
      "0": "2",
      "group": "num",
    },
  ],
]
`;

exports[`Types arguments range i32..f32, default f32 = f32 1`] = `
"(func $f 
  (export \\"f\\") 
  (param $x f32) 
  (result f32) 
  (local.get $x))"
`;

exports[`Types arguments range i32..f32, default f32 = f32 2`] = `
Array [
  "f32.convert_i32_u",
  Array [
    "i32.const",
    Token {
      "0": "1",
      "group": "num",
    },
  ],
]
`;

exports[`Types arguments range i32..f32, default f32 = f32 3`] = `
Array [
  "f32.const",
  Token {
    "0": "2",
    "1": ".",
    "2": "0",
    "group": "num",
  },
]
`;

exports[`Types arguments range f32..f32, default i32 = f32 1`] = `
"(func $f 
  (export \\"f\\") 
  (param $x f32) 
  (result f32) 
  (local.get $x))"
`;

exports[`Types arguments range f32..f32, default i32 = f32 2`] = `
Array [
  "f32.const",
  Token {
    "0": "1",
    "1": ".",
    "2": "0",
    "group": "num",
  },
]
`;

exports[`Types arguments range f32..f32, default i32 = f32 3`] = `
Array [
  "f32.const",
  Token {
    "0": "2",
    "1": ".",
    "2": "0",
    "group": "num",
  },
]
`;

exports[`Types arguments range f32..f32, default i32 = f32 4`] = `
Array [
  "f32.convert_i32_s",
  Array [
    "i32.const",
    Token {
      "0": "3",
      "group": "num",
    },
  ],
]
`;

exports[`Types arguments range i32..i32, default i32 = i32 1`] = `
"(func $f 
  (export \\"f\\") 
  (param $x i32) 
  (result i32) 
  (local.get $x))"
`;

exports[`Types arguments range i32..i32, default i32 = i32 2`] = `
Array [
  "i32.const",
  Token {
    "0": "2",
    "group": "num",
  },
]
`;

exports[`Types arguments range i32..i32, default i32 = i32 3`] = `
Array [
  "i32.const",
  Token {
    "0": "5",
    "group": "num",
  },
]
`;

exports[`Types arguments range i32..i32, default i32 = i32 4`] = `
Array [
  "i32.const",
  Token {
    "0": "3",
    "group": "num",
  },
]
`;

exports[`Types arguments use default i32 1`] = `
"(func $f 
  (export \\"f\\") 
  (result i32) 
  (call $a 
    (i32.const 5)))"
`;

exports[`Types arguments use default exported i32 1`] = `"any"`;

exports[`Types arguments use default exported i32 2`] = `
"(func $__start__ 
  (export \\"__start__\\") 
  (call $__begin__) 
  (global.set $export/a/x 
    (i32.const 5)))"
`;

exports[`Types arguments use default exported i32 3`] = `
"(func $f 
  (export \\"f\\") 
  (result i32) 
  (call $a 
    (global.get $export/a/x)))"
`;

exports[`Types arguments use default f32 1`] = `
"(func $f 
  (export \\"f\\") 
  (result f32) 
  (call $a 
    (f32.const 5.5)))"
`;

exports[`Types arguments default to range i32 1`] = `
"(func $f 
  (export \\"f\\") 
  (result i32) 
  (call $a 
    (i32.const 1)))"
`;

exports[`Types arguments default to range bool 1`] = `
"(func $f 
  (export \\"f\\") 
  (result i32) 
  (call $a 
    (i32.const 0)))"
`;

exports[`Types arguments default to range f32 1`] = `
"(func $f 
  (export \\"f\\") 
  (result f32) 
  (call $a 
    (f32.const 0.5)))"
`;
